devito/types/grid.py

```diff
@@ -421,3 +425,125 @@ class Interior(SubDomain):
+    >>> xM = np.array([Nx/2+1, 1], dtype=np.int32)
+
+    Along a dimension where all bounds are the same we can use the
+    following shorthand:
+
+    >>> ym = 1 # which is equivalent to 'np.array([1, 1], dtype=np.int32)'
+    >>> yM = 1
+
+    Combine the data into the required form:
+
+    >>> bounds = (xm, xM, ym, yM)
+
+    Create our set of subdomains passing the number of domains and the
+    bounds:
+
+    >>> my_sd = MySubdomains(N=n_domains, bounds=bounds)
+
+    Create a grid and iterate a function within the define subdomains:
+
+    >>> grid = Grid(extent=(Nx, Ny), shape=(Nx, Ny), subdomains=(my_sd, ))
+    >>> f = Function(name='f', grid=grid, dtype=np.int32)
+    >>> eq = Eq(f, f+1, subdomain=grid.subdomains['mydomains'])
+    >>> op = Operator(eq)
+    >>> summary = op.apply()
+    >>> f.data
+    Data([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+          [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=int32)
+    """
+
+    def __init__(self, **kwargs):
+        super(SubDomainSet, self).__init__()
+        self._n_domains = kwargs.get('N', 1)
+        self._bounds = kwargs.get('bounds', None)
+
+    def __subdomain_finalize__(self, dimensions, shape):
+        super(SubDomainSet, self).__subdomain_finalize__(dimensions, shape)
+        n = Dimension(name='n')
+        self._implicit_dimension = n
+
+    @property
+    def n_domains(self):
+        return self._n_domains
+
+    @property
+    def bounds(self):
+        return self._bounds
+
+    @cached_property
+    def _implicit_eq_dat(self):
+        if not len(self._bounds) == 2*len(self.dimensions):
+            raise ValueError("Left and right bounds must be supplied for each dimension")
+        n_domains = self.n_domains
+        i_dim = self._implicit_dimension
+        b_f = {}
+        dat = []
+        # Organise the data contained in 'bounds' into a form such that the
+        # associated implicit equations can easily be created.
+        for j in range(0, len(self._bounds)):
+            index = floor(j/2)
+            d = self.dimensions[index]
+            if j % 2 == 0:
+                fname = d.min_name
+            else:
+                fname = d.max_name
+            b_f[fname] = Function(name=fname, shape=(n_domains, ),
+                                  dimensions=(i_dim, ), dtype=np.int32)
+            if isinstance(self._bounds[j], int):
+                bounds = np.full((n_domains,), self._bounds[j], dtype=np.int32)
+                b_f[fname].data[:] = bounds
+            else:
+                b_f[fname].data[:] = self._bounds[j]
+            dat.append({'rhs': d.thickness[j % 2][0], 'lhs': b_f[fname][i_dim]})
```

Comment:
I suppose you want a ``namedtuple`` here, not a ``dict``

Reply:
Or you could alternatively return an `Eq`, and then in `operator.py` you could rebuild it (`eq.func(*eq.args, ...)`) to append subdomain and implicit_dims